File: Binary Search/binary_search_recursive.go
/*
	Implelemtation of Recursive Binary Search

	In computer science, binary search, also known as half-interval search,
	logarithmic search, or binary chop, is a search algorithm that
	finds the position of a target value within a sorted array.
	Binary search compares the target value to the middle element of the array.
	If they are not equal, the half in which the target cannot lie is eliminated
	and the search continues on the remaining half,
	again taking the middle element to compare to the target value,
	and repeating this until the target value is found.
	If the search ends with the remaining half being empty, the target is not in the array.
	Source(https://en.wikipedia.org/wiki/Binary_search_algorithm)
*/
package main

import "fmt"

func BinarySearchRecursive(Arr []int, key int) bool {
	low := 0
	high := len(Arr) - 1
	if low <= high {
		mid := low + (high - low) / 2 // prevent overflow
		if Arr[mid] > key { // Element in mid is higher than target.
			return BinarySearchRecursive(Arr[:mid], key) // search from start to mid 
		} else if Arr[mid] < key { // Element in mid is lower than target.
			return BinarySearchRecursive(Arr[mid + 1:], key) // search from mid + 1 to last
		} else {
			return true
		}
	}
	return false
}

func main() {
	Arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	fmt.Println(BinarySearchRecursive(Arr, 5))
	fmt.Println(BinarySearchRecursive(Arr, 6))
	fmt.Println(BinarySearchRecursive(Arr, -1))
	fmt.Println(BinarySearchRecursive(Arr, -10))
	fmt.Println(BinarySearchRecursive(Arr, 70))
}

File: Binary Search/BinarySearchRecursive.java

/*
 * Problem : Implement Binary Search using Recursion.
 * 
 * RECURSION : Function that calls itself is called recursion.
 * 
 * Algorithm:
 * 1) Get an sorted array and a target to be found in the array.
 * 2) Find the middle index by dividing the first and last element.
 * 3) If the middle element is equal to target , return the mid index.
 * 4) If the middle element is greater than the target , in the next recursive call make last element as mid-1
 * 5) If the middle element is smaller than the target , in the next recursive call make first element as mid+1
 * 6) If start is less than end ,ie,if the element not found , return -1.
 * 
 * Time complexity: worst case: O(logN) , N - size of the array.
 * 
 * Space Complexity: O(logN) (since we use stack for function calls)
 * 
 */


public class BinarySearchRecursive {

    public static void main(String[] args) {
        int[] arr = {1,2,5,89,99,101};
        int target = 1;
        System.out.println(binarySearch(arr , target,0 , arr.length-1));
    }

    static int binarySearch(int[] arr ,int target, int start , int end) {

       int mid = (start + end)/2;

       if (start > end) {
           return -1;
       }
       if (arr[mid] == target) {
           return mid;
       }

       else if (arr[mid] > target) {
           return binarySearch(arr , target , start , mid-1);
       }

       return  binarySearch(arr ,target , mid+1, end);

    }
}
